Support opt-out of extra derived traits for rejections for `#[derive(FromRequest)]`

davidpdrsn opened the pull request:
Makes it possible to write

```rust
#[derive(FromRequest)]
#[from_request(rejection_derive(!Display, !Error))]
struct Extractor {
    other: OtherExtractor,
}
```

Which opts out of the rejection implementing `Display` and `Error`.

The allowed syntax inside `rejection_derive` is intentionally quite limited. Can always be expanded in the future if we needed.

Fixes https://github.com/tokio-rs/axum/issues/726
Fixes https://github.com/tokio-rs/axum/issues/725

davidpdrsn commented on axum-macros/src/from_request/attr.rs:
```diff
@@ -0,0 +1,202 @@
+            Err(lh.error())
+        }
+    }
+}
```
The duplication in this parsing code bothers me a little but I couldn't find a nicer solution that gave good error messages.

jplatte replied:
Seems fine to me as-is.

jplatte commented on axum-macros/src/from_request.rs:
```diff
@@ -270,7 +288,7 @@ fn extract_each_field_rejection(
     };
 
-    let impl_display = {
+    let impl_display = if rejection_derive_opt_outs.derive_display() {
```
Could use `bool::then` (plus `.transpose()?`) here.

davidpdrsn replied:
Hm maybe. I think its fine as is.

jplatte commented on axum-macros/src/from_request/attr.rs:
```diff
@@ -0,0 +1,202 @@
+
+            let lh = input.lookahead1();
+            if lh.peek(kw::Debug) {
+                debug = Some(input.parse()?);
```
Should probably raise an error on multiple assignment here too? Could avoid the extra nesting by moving the assignment or error into the helper function.

davidpdrsn commented:
I fixed a few other things I discovered:

- Added docs for `rejection_derive`
- Multiple opt outs of the same trait like you mentioned
- Using both `via` and `rejection_derive`. Since `via` on the container changes it to use the "via extractors"'s rejection, and doesn't generate a few type, you cannot opt out of which traits are derived for it
- Better error message if opting out of `Debug` or `Display` with also opting out of `Error`. `std::error::Error` requires `Debug` and `Display` so the error message was slightly confusing

jplatte commented on axum-macros/src/from_request.rs:
```diff
@@ -313,11 +333,19 @@ fn extract_each_field_rejection(
+        None
+    };
+
+    let impl_debug = if rejection_derive_opt_outs.derive_debug() {
+        Some(quote! { #[derive(Debug)] })
+    } else {
+        None
     };
```
Here you don't even need `.transpose()?` 😉

```suggestion
    let impl_debug = rejection_derive_opt_outs.derive_debug().then(|| {
        quote! { #[derive(Debug)] }
    });
```

But of course the current code is fine too.

davidpdrsn replied:
Yeah thats definitely nicer!
